Divergence and Gradient integrators as well as PA versions

jandrej opened the pull request:
This PR introduces the divergence and gradient bilinear integrators. It introduces facilities to apply boundary conditions to general rectangular operators. These changes required a lot of additions to hypre functions, but do not change any existing interfaces.

For PA it was necessary to introduce a `MixedBilinearFormExtension`.

Some application kernels contain shared memory version from @camierjs.

It also includes unit tests for the PA version of the operators.

Resolves #1158.

- [x] Code builds.
- [x] Code passes make style.
- [x] Update CHANGELOG.

| PR | Author | Editor | Reviewers  | Assignment | Approval | Merge |
| --- | --- | --- | ---  | --- | --- | --- |
| https://github.com/mfem/mfem/pull/1164 | @jandrej | @tzanio | @mlstowell + @jamiebramwell | 12/2/19 | 01/08/20 | ⌛due 01/20/20 |

tzanio commented on fem/bilinearform.hpp:
```diff
@@ -525,6 +531,9 @@ class BilinearForm : public Matrix
    void SetDiagonalPolicy(DiagonalPolicy policy);
 
+   /// Indicate that integrators are not owned by the BilinearForm
+   void SetExternBFS(int e = 1) { extern_bfs = e; };
```
I will suggest a move descriptive name and dropping the parameter, e.g. `void UseExternalIntegrators()'.

pazner replied:
+1 for a more descriptive name

pazner commented on fem/bilinearform.cpp:
```diff
@@ -1470,17 +1561,91 @@ void MixedBilinearForm::EliminateTestDofs (const Array<int> &bdr_attr_is_ess)
+
+   mat->Finalize();
+
+   if (test_P) // TODO: Must actually check for trial_P too
```
This TODO should be addressed?

jandrej replied:
I'm not 100% sure what we should do here.

pazner commented on fem/bilinearform.cpp:
```diff
@@ -1470,17 +1561,91 @@ void MixedBilinearForm::EliminateTestDofs (const Array<int> &bdr_attr_is_ess)
+
+   mat_e = new SparseMatrix(mat->Height(), mat->Width());
+   mat->EliminateCols(ess_trial_tdof_marker, *mat_e);
+   // TODO: WP: are we doing the right thing here?
```
This TODO can be removed

pazner commented on fem/bilinearform_ext.hpp:
```diff
@@ -134,6 +136,120 @@ class MFBilinearFormExtension : public BilinearFormExtension
+   void MultTranspose(const Vector &x, Vector &y) const {}
+   void Update() {}
+   ~FAMixedBilinearFormExtension() {}
+};
```
Should we hold off on this FA stuff? It's more-or-less a placeholder now

pazner commented on fem/bilininteg.cpp:
```diff
@@ -1925,7 +1998,7 @@ void VectorFEMassIntegrator::AssembleElementMatrix2(
 }
 
-void VectorDivergenceIntegrator::AssembleElementMatrix2(
+void DivergenceIntegrator::AssembleElementMatrix2(
```
This name change will break backwards compatibility (although I agree it's a better name)

pazner commented on linalg/operator.cpp:
```diff
@@ -336,4 +412,76 @@ void ConstrainedOperator::Mult(const Vector &x, Vector &y) const
+   mem_class = A->GetMemoryClass()*Device::GetMemoryClass();
+   MemoryType mem_type = GetMemoryType(mem_class);
+   trial_list.Read(); // TODO: just ensure 'list' is registered, no need to copy it
+   test_list.Read(); // TODO: just ensure 'list' is registered, no need to copy it
```
Have these TODOs been addressed?

pazner commented on linalg/operator.hpp:
```diff
@@ -30,7 +31,21 @@ class Operator
+   void InitTVectors(const Operator *Po, const Operator *Ri,
+                     Vector &x, Vector &b,
+                     Vector &X, Vector &B) const;
+
```
Should this be in the public interface? Maybe it can be moved to be private.

pazner commented on fem/bilininteg_divergence.cpp:
```diff
@@ -0,0 +1,1073 @@
+      //double (*QQD2)[MQ1][MD1] = (double (*)[MQ1][MD1]) (sm1+2);
+      double (*QDD0)[MD1][MD1] = (double (*)[MD1][MD1]) (sm0+0);
+      //double (*QDD1)[MD1][MD1] = (double (*)[MD1][MD1]) (sm0+1);
+      //double (*QDD2)[MD1][MD1] = (double (*)[MD1][MD1]) (sm0+2);
```
These commented-out lines can be removed

pazner commented on fem/bilininteg_gradient.cpp:
```diff
@@ -0,0 +1,800 @@
+{
+   // TODO
+   MFEM_ASSERT(false, "SHARED MEM NOT PROGRAMMED YET");
+}
```
Better to remove the placeholder for the shared memory version until it's implemented

pazner commented on fem/bilininteg_gradient.cpp:
```diff
@@ -0,0 +1,800 @@
+                                       int q1d = 0)
+{
+   // TODO
+   MFEM_ASSERT(false, "Gradient PA Apply Transpose 3D NOT PROGRAMMED YET");
```
Nicer error message? e.g. `"GradientPAApplyTranspose 3D not implemented"`

pazner commented on fem/bilininteg_gradient.cpp:
```diff
@@ -0,0 +1,800 @@
+      else
+      {
+         return PAGradientApply3D(NE,B,G,Bt,op,x,y,TR_D1D,TE_D1D,Q1D);
+      }
```
Why do we check for transpose in the `dim=3` case but not `dim=2`? And is the `transpose` flag ever set given that `GradientIntegrator::AddMultTransposePA` (defined below) throws an error?

pazner commented on fem/bilininteg_divergence.cpp:
```diff
@@ -0,0 +1,1073 @@
+   if (dim == 3)
+   {
+      return PADivergenceApply3D(NE,B,G,Bt,op,x,y,TR_D1D,TE_D1D,Q1D);
+   }
```
@tzanio @v-dobrev @camierjs We need to figure out a way of doing template specialization for common cases for these kernels without hugely increasing compile times. Here we're only using the default fallback kernel because otherwise (now that we're adding more and more kernels) the compilation is just too long. Also the `switch ((D1D << 4 ) | Q1D)` approach is error-prone and requires too much manual intervention (e.g. picking exactly the right Q1D to match the quadrature rule).

jamiebramwell commented on linalg/hypre_parcsr.hpp:
```diff
@@ -45,7 +45,14 @@ void hypre_ParCSRMatrixEliminateAXB(hypre_ParCSRMatrix *A,
+
+/** Eliminate rows from a hypre ParCSRMatrix, setting all entries in the listed
+    rows of the matrix to zero. */
+void hypre_ParCSRMatrixEliminateRows(hypre_ParCSRMatrix *A,
```
This goes beyond this PR, but has anyone looked at upstreaming these functions to HYPRE? Any thoughts @tzanio ?

tzanio replied:
This is on my todo list...

jamiebramwell commented on linalg/hypre.hpp:
```diff
@@ -525,6 +525,13 @@ class HypreParMatrix : public Operator
+       matrix and Ae sum to the original matrix. */
+   HypreParMatrix* EliminateCols(const Array<int> &cols);
+
+   void EliminateRows(const Array<int> &rows);
```
This is missing a function description. It's pretty obvious what it does, but I think it's nice to include for completeness.

jamiebramwell commented on fem/bilininteg_gradient.cpp:
```diff
@@ -0,0 +1,775 @@
+// Copyright (c) 2019, Lawrence Livermore National Security, LLC. Produced at
```
While this is not unique to this PR, these kernels are very hard to read without any comments. As someone who will probably have to write specialized kernels in the future, comments outlining the general method for each operator would be greatly appreciated.

mlstowell replied:
I have to second this.  Since these files don't have headers we should put detailed, header-level, documentation here before each function.  At a minimum we should have descriptions of the arguments but, as @jamiebramwell suggests, an overview of how the function works would be ideal.

mlstowell commented on tests/unit/fem/test_pa_kernels.cpp:
```diff
@@ -0,0 +1,217 @@
+   return field2.Norml2();
+}
+
+TEST_CASE("PA Divergence", "[PartialAssembly], [MixedBilinearFormBC]")
```
I like your tests but I'm curious about "BC" in "MixedBilinearFormBC".  What does this signify?

jandrej replied:
This is cruft that I didn't clean up. I wanted to test the boundary condition application as well but didn't complete it in the commit.

mlstowell commented on fem/bilinearform.hpp:
```diff
@@ -676,8 +689,25 @@ class MixedBilinearForm : public Matrix
+   virtual const Operator *GetRestriction() const
+   { return trial_fes->GetRestrictionMatrix(); }
+   /// Get the output finite element space prolongation matrix
+   virtual const Operator *GetOutputProlongation() const
```
Since this is an FEM class rather than a linear algebra class perhaps it would be better to use "Trial" and "Test" rather than "" and "Output" in these method names.  Personally, I tend to get "trial" and "test" confused but they are part of the convention.

mlstowell commented on fem/bilinearform.cpp:
```diff
@@ -960,6 +958,18 @@ void BilinearForm::EliminateVDofsInRHS(
    mat->PartMult(vdofs, x, b);
 }
 
+void BilinearForm::Mult(const Vector &x, Vector &y) const
```
Don't we need similar changes to `BilinearForm::AddMult()` to support `ext`?

jandrej replied:
Do you want me to add them? We did this in the Mixed case because they were needed for some operations.

mlstowell replied:
I would suggests adding them because it doesn't seem like much effort.  However, I won't object if you postpone this and start an issue so we won't forget to address it.

jandrej replied:
I'd prefer to postpone this to a different PR.

mlstowell commented on fem/bilinearform.hpp:
```diff
@@ -676,8 +689,25 @@ class MixedBilinearForm : public Matrix
 
    void operator=(const double a) { *mat = a; }
```
I'm a little concerned about methods like this, and the `SpMat` methods earlier, that will fail if partial assembly was chosen.  Certainly for the `SpMat` methods we could make the argument that responsible developers should never call them if they requested that matrices not be assembled but we have no control over this and it can easily lead to runtime errors.  I prefer compile time errors whenever possible.

I wonder what people's reaction would be if I suggested that we should have an abstract base class for bilinear forms which contained a minimal interface and then two or more derived classes for concrete implementations of fully assembled and partially assembled bilinear forms.  Each concrete class could then have separate APIs containing only those methods which could actually be implemented.

jandrej replied:
Do you think this should be addressed in this PR or discussed in the meeting for another PR?

mlstowell replied:
A meeting would probably be best.

mlstowell commented on fem/bilininteg_gradient.cpp:
```diff
@@ -0,0 +1,775 @@
+}
+
+} // namespace mfem
+
```
These kernel files are very cleanly written!  Adding comment blocks as described earlier would make them perfect.

jandrej commented:
I addressed all comments either with changes or replies. I did not address the comments in the kernels, I completely understand the issue with writing them and implementations can lead as an example for people who want customized kernels. As stated in other PR's we have to think harder about how to reduce the labor for this.

mlstowell commented on fem/bilinearform.hpp:
```diff
@@ -325,6 +325,12 @@ class BilinearForm : public Matrix
+   virtual const Operator *GetOutputProlongation() const
+   { return GetProlongation(); }
+   /// Get the output finite element space restriction matrix
+   virtual const Operator *GetOutputRestriction() const
```
Should change these from "Output" to "Test" as well or eliminate them.

jandrej replied:
Removed.

tzanio commented:
Reminder: the reviews for this PR are due in less than a week

cc: @mlstowell, @jamiebramwell

tzanio commented:
@jamiebramwell were you comments addressed, and if so, can you formally approve?

tzanio commented on tests/unit/fem/test_pa_kernels.cpp:
```diff
@@ -0,0 +1,227 @@
+#include "catch.hpp"
```
Please add the Copyright header in the beginning of this file (as in the other tests).

tzanio commented:
@jandrej, before we merge this into `next`, can you please do the following:
- [ ] Update `CHANGELOG`
- [ ] Go through [all the comments](https://github.com/mfem/mfem/pull/1164/files), and either resolve them (if fixed) or address them (if not)
- [ ] Stop by to discuss in person if there are questions :)

tzanio commented:
Merged in `next` for testing ...

v-dobrev commented on fem/bilininteg.hpp:
```diff
@@ -43,6 +43,9 @@ class BilinearFormIntegrator : public NonlinearFormIntegrator
    virtual void AssemblePA(const FiniteElementSpace &fes);
+   /** Used with BilinearFormIntegrators that have different spaces. */
+   virtual void AssemblePA(const FiniteElementSpace &trial_fes,
+                           const FiniteElementSpace &test_fes);
```
While building the `next` branch with CUDA enabled, I noticed that this new overload of `AssemblePA` causes numerous warnings like this:
```console
fem/bilininteg.hpp(1671): warning: overloaded virtual function "mfem::BilinearFormIntegrator::AssemblePA" is only partially overridden in class "mfem::GradientIntegrator"
```
One way to fix this is to add `using BilinearFormIntegrator::AssemblePA;` to all subclasses. Another is to change the name of the method slightly, e.g. `MixedAssemblePA`.

v-dobrev replied:
I see that there was a similar issue in #1159 that @camierjs fixed there by adding `using` statements. In that PR the same method as above was introduced at the level of the parent class, `NonlinearFormIntegrator`.

jandrej replied:
I don't have a cuda compiler setup. Can someone who has make the appropriate changes?

tzanio replied:
@camierjs, do you think you can help @jandrej with this?

camierjs replied:
`done`

jandrej replied:
Thanks!

tzanio commented:
Re-merged in `next` for testing ...

tzanio commented:
@jandrej, can you resolve the conflicts?

tzanio commented:
Added few more days of testing due to changes...

jandrej commented:
> @jandrej, can you resolve the conflicts?

Done.

tzanio commented:
Re-merged in `next` for testing ...

ben-s-southworth commented:
From https://github.com/mfem/mfem/issues/1158 and originally https://github.com/mfem/mfem/issues/793, I don't see anything here that addresses the HypreParMatrix ownership? This comes up more often for me than the integrators, I need a matrix, but not the integrators or bilinear forms.

v-dobrev commented on fem/bilinearform.hpp:
```diff
@@ -533,6 +539,9 @@ class BilinearForm : public Matrix
    void SetDiagonalPolicy(DiagonalPolicy policy);
 
+   /// Indicate that integrators are not owned by the BilinearForm
+   void UseExternalIntegrators() { extern_bfs = 1; };
```
@ben-s-southworth, I think this is the method you are looking for.

v-dobrev replied:
At least with respect to #1158.

ben-s-southworth replied:
@v-dobrev Yep, https://github.com/mfem/mfem/issues/1158 is addressed here, but I don't think anything addresses the HypreParMatrices in https://github.com/mfem/mfem/issues/793, unless I missed it

v-dobrev replied:
My understanding is that the discussion in #793 is still open. I don't think it was moved to #1158 -- at least that is my understanding.

In any case, I agree that this PR does not address #793.

tzanio replied:
Correct, the discussion in #793 is outstanding and not addressed here.

v-dobrev commented on linalg/sparsemat.hpp:
```diff
@@ -424,7 +424,7 @@ class SparseMatrix : public AbstractSparseMatrix
    bool Finalized() const { return !A.Empty(); }
    /// Returns whether or not the columns are sorted.
-   bool areColumnsSorted() const { return isSorted; }
+   bool ColumnsAreSorted() const { return isSorted; }
```
@tzanio, please do not sneak in interface changes like this -- just because you didn't like the name, and right before merging in master.

tzanio replied:
This was an editorial decision that I discussed with @jandrej in person. It was not a question of taste, but of conformity with the current naming practice in MFEM.
